pkg/istio/rbac/v2/provider.go

```diff
@@ -147,8 +96,65 @@ func (p *v2) ConvertAthenzModelIntoIstioRbac(athenzModel athenz.Model, serviceNa
 				to.Operation.Paths = []string{path}
 			}
 			rule.To = append(rule.To, to)
-			rules = append(rules, rule)
 		}
+
+		// group by role, for each role, form rule_from from role members,
+		// skip if rule.To is nil, indicating no assertion match with service
+		if rule.To == nil {
+			continue
+		}
+		from_principal := &v1beta1.Rule_From{
+			Source: &v1beta1.Source{},
+		}
+		from_requestPrincipal := &v1beta1.Rule_From{
+			Source: &v1beta1.Source{},
+		}
+		// role name should match zms resource name
+		for _, roleMember := range athenzModel.Members[role] {
+			res, err := common.CheckAthenzMemberExpiry(roleMember)
+			if err != nil {
+				log.Errorf("error when checking athenz member expiration date, skipping current member: %s, error: %s", roleMember.MemberName, err)
+				continue
+			}
+			if !res {
+				log.Infoln("member expired, skip adding member to authz policy resource, member: ", roleMember.MemberName)
+				continue
+			}
+			res, err = common.CheckAthenzSystemDisabled(roleMember)
+			if err != nil {
+				log.Errorf("error when checking athenz member system disabled, skipping current member: %s, error: %s", roleMember.MemberName, err)
+				continue
+			}
+			if !res {
+				log.Infoln("member expired, skip adding member to authz policy resource, member: ", roleMember.MemberName)
+				continue
+			}
+
+			spiffeName, err := common.MemberToSpiffe(roleMember)
+			if err != nil {
+				log.Errorln("error converting role member to spiffeName: ", err.Error())
+				continue
+			}
+			from_principal.Source.Principals = append(from_principal.Source.Principals, spiffeName)
+			if p.enableOriginJwtSubject {
+				originJwtName, err := common.MemberToOriginJwtSubject(roleMember)
+				if err != nil {
+					log.Errorln(err.Error())
+					continue
+				}
+				from_requestPrincipal.Source.RequestPrincipals = append(from_requestPrincipal.Source.RequestPrincipals, originJwtName)
+			}
+		}
+		//add role spiffe for role certificate
+		roleSpiffeName, err := common.RoleToSpiffe(string(athenzModel.Name), string(role))
+		if err != nil {
+			log.Errorln("error when convert role to spiffe name: ", err.Error())
+			continue
+		}
+		from_principal.Source.Principals = append(from_principal.Source.Principals, roleSpiffeName)
+		rule.From = append(rule.From, from_principal)
+		rule.From = append(rule.From, from_requestPrincipal)
```

Comment:
Should we append this only if enableOriginJwtSubject is enabled? Not sure how an empty array will be treated here if the flag is off.

Reply:
makes sense